server/http/helloapp/oauth.go

```diff
@@ -0,0 +1,128 @@
+package helloapp
+
+import (
+	"encoding/json"
+	"net/http"
+
+	"github.com/mattermost/mattermost-plugin-api/experimental/bot/logger"
+	"github.com/mattermost/mattermost-plugin-api/experimental/oauther"
+	"github.com/mattermost/mattermost-plugin-apps/server/apps"
+	"github.com/mattermost/mattermost-plugin-apps/server/constants"
+	"github.com/mattermost/mattermost-server/v5/model"
+	"github.com/pkg/errors"
+	"golang.org/x/oauth2"
+)
+
+func (h *helloapp) InitOAuther() error {
+	oauth2Config, err := h.GetOAuthConfig()
+	if err != nil {
+		return err
+	}
+	h.OAuther = oauther.NewFromClient(h.apps.Mattermost,
+		*oauth2Config,
+		h.finishOAuth2Connect,
+		logger.NewNilLogger(),
+		oauther.OAuthURL(constants.HelloAppPath+PathOAuth2),
+		oauther.StorePrefix("hello_oauth_"))
+	return nil
+}
+
+func (h *helloapp) handleOAuth(w http.ResponseWriter, req *http.Request) {
+	if h.OAuther == nil {
+		http.Error(w, "OAuth not initialized", http.StatusInternalServerError)
+		return
+	}
+	h.OAuther.ServeHTTP(w, req)
+}
+
+func (h *helloapp) GetOAuthConfig() (*oauth2.Config, error) {
+	conf := h.apps.Configurator.GetConfig()
+
+	id, secret, err := h.getOAuth2AppCredentials()
+	if err != nil {
+		return nil, errors.Wrap(err, "failed to retrieve App OAuth2 credentials")
+	}
+
+	return &oauth2.Config{
+		ClientID:     id,
+		ClientSecret: secret,
+		Endpoint: oauth2.Endpoint{
+			AuthURL:  conf.MattermostSiteURL + "/oauth/authorize",
+			TokenURL: conf.MattermostSiteURL + "/oauth/access_token",
+		},
+		RedirectURL: h.AppURL(PathOAuth2Complete),
+		// TODO Scopes:
+	}, nil
+}
+
+func (h *helloapp) startOAuth2Connect(userID string, callOnComplete apps.Call) (string, error) {
+	state, err := json.Marshal(callOnComplete)
+	if err != nil {
+		return "", err
+	}
+
+	err = h.OAuther.AddPayload(userID, state)
+	if err != nil {
+		return "", err
+	}
+	return h.OAuther.GetConnectURL(), nil
+}
+
+func (h *helloapp) finishOAuth2Connect(userID string, token oauth2.Token, payload []byte) {
+	call := apps.Call{}
+	err := json.Unmarshal(payload, &call)
+	if err != nil {
+		return
+	}
+
+	// TODO 2/5 we should wrap the OAuther for the users as a "service" so that
+	//  - startOAuth2Connect is a Call
+	//  - payload for finish should be a Call
+	//  - a Wish can check the presence of the acting user's OAuth2 token, and
+	//    return Call startOAuth2Connect(itself)
+	// for now hacking access to apps object and issuing the call from within
+	// the app.
```

Comment:
This makes sense. It would alleviate the needs of the App in terms of storage (it no longer has to store the tokens, just ask for the token, and if it doesn't receive it, understand it is not authorized by that user).

This will also make things slightly more secure, since the app information stays in mattermost instead of residing in the app.

The only problem is then how does the App identifies itself to ask for the tokens? Are we sharing yet another secret?

Reply:
I think the App should _store_ the tokens and the app credentials, as it does now. I just want the wire the flow of InstallWish -> ConnectWish -> ...link/browser... -> CompleteCallbackWish (aka connected install)